### ISSUE TITLE
Question: fb2 in-page footnotes justified

### ISSUE THREAD

**Issue by hius07** [koreader/koreader#6344]
https://github.com/poire-z, can you please help me to have fb2 in-page footnotes justified, same as popup footnote.
I've played with User style tweaks, cr3.ini, footnotewidget.lua with no success.
![Reader_2020-07-03_062817](https://user-images.githubusercontent.com/62179190/86462203-e6b7ac00-bd33-11ea-9297-51b4d07f34f6.png)
![Reader_2020-07-03_062826](https://user-images.githubusercontent.com/62179190/86462226-f20ad780-bd33-11ea-9e84-92d6093a8101.png)

**Comment by poire-z** [comment]
A user style tweak with this should work:
```css
body[name="notes"] section,
body[name="comments"] section {
   text-align: justify !important;
}
```
Does that work ?
If not, can you try to see by inspecting the HTML where a `text-align: left` comes from.

I added recently `text-align: justify;` to the "book style tweak" sample:
https://github.com/koreader/koreader/blob/8a61e70fadb74189b1196ac9a7d07e49f6c0111a/frontend/apps/reader/modules/readerstyletweak.lua#L591-L602
Because I sometimes (but rarely) got them non-justified, and I figured it would not hurt there (a user can just erase that sample line).
Dunno if we should switch all in-page footnotes tweaks to `text-align: justify`.

**Comment by poire-z** [comment]
Even if it works, can you try to see where a text-align: left comes from. Obviously, it's not you who set it via another style tweak.

**Comment by hius07** [comment]
The user style tweak doesn't work.
I've checked cr3.ini, fb2.css, cr3.css, the book itself - no left-aligned instructions.
The text of the footnote is left-aligned both in the in-page footnote and in the end of the book.

**Comment by hius07** [comment]
I've applied the sample of Book-specific tweak - doesn't work.

**Comment by hius07** [comment]
I cannot trace the source of the text-align:left instruction, the html code shows just <body name="notes">, <section> and <p>.

**Comment by poire-z** [comment]
Try this, forcing it on all descendants:
```css
body[name="notes"] section *,
body[name="comments"] section * {
   text-align: justify !important;
}
```

It's only on this specific FB2 book (can you share it?), or does it happen with all FB2 books?

**Comment by hius07** [comment]
Doesn't work. All books.

**Comment by poire-z** [comment]
So, please share one (that you can share), so we can see if it's you only - and if I can reproduce I'll investigate.

Btw, on your first post, it seems fine on the 2nd screenshot (?) - or did I not undestand the issue?

**Comment by hius07** [comment]
The second screenshot shows popup footnote.
[Лондон - Мартин Иден.fb2.zip](https://github.com/koreader/koreader/files/4869647/-.fb2.zip)

**Comment by poire-z** [comment]
> The second screenshot shows popup footnote

Oh, right :)

OK, I can reproduce it - and it does not happen with a koreader 2020.04.1-12 I had around.
So, probably a bug caused by some of my recent changes... :(
I'll look into it - you can stop fighting it with style tweaks.

**Comment by poire-z** [comment]
Cause by https://github.com/koreader/crengine/commit/28ae22c8e311d811be2b1923d292619cc51faf8f .

**Comment by hius07** [comment]
I've changed in fb2.css run-in to auto and got justified footnote, the text starting the next line of the footnote number. User style tweak works.

**Comment by poire-z** [comment]
Should be fixed with https://github.com/koreader/crengine/pull/353.
(You'll find there an alternative way of handling FB2 footnotes with floats, to replace the similar stuff in fb2.css with - if you need a quick temporary fix :)

**Comment by hius07** [comment]
Thank you very much, I'll wait for the build to test.

**Comment by hius07** [comment]
> Dunno if we should switch all in-page footnotes tweaks to `text-align: justify`.

Can in-page footnotes follow the alignment settings of the main text?

**Comment by poire-z** [comment]
For FB2, they did, and they will with the fix.
They are a `<p>` , so are affected by the `Style tweaks > Text > Alignment > Left/justify most text`.
And FB2 do not have `class=` I think, and aren't really styled by publishers (is this right ?)

With EPUBs, it's more wild, as they can be in a DIV, and some publisher can explicitely set them to `text-align: left`. But some people may like them like that. I just wondered how many people :)
It's quite rare - so I guess we can let them as-is, not overriding publishers and using our default of justify as a fallback.

**Comment by hius07** [comment]
Nice, in fb2 books by default we will have both main text and in-page footnotes justified (as per fb2.css setting for `<p>`).
Yes, fb2 books are not really styled by publishers, just marked up.

**Comment by hius07** [comment]
In-page footnotes are now perfectly justified, thanks a lot!

I'm not sure if it is worth your attention, anyway: the space between the number and the text in in-page footnotes is doubled comparing to popup footnotes.

![5](https://user-images.githubusercontent.com/62179190/87006524-7b555a80-c1c9-11ea-937f-f1dd78235750.jpg)

**Comment by poire-z** [comment]
That is (and has been for a long time - it's original upstream code) explicitely done by adding 2 UNICODE_NO_BREAK_SPACE between the number and the text:
https://github.com/koreader/crengine/blob/59483446bc777b211cb23181bc8f6c60b94a1c1b/crengine/src/lvrend.cpp#L3175-L3184
and some checks to avoid having these spaces expanded or compressed by text justification:
https://github.com/koreader/crengine/blob/59483446bc777b211cb23181bc8f6c60b94a1c1b/crengine/src/lvtextfm.cpp#L3052-L3062

So, I guess someone prefered them that way, unless it's just some technical trick, not sure.
About that last bit, I added this as a comment in some current work on cleaning some stuff:
```
// For FB2 footnotes, renderFinalBlock() has added between the
// footnote number (set display: run-in) and the following text
// these 2 UNICODE_NO_BREAK_SPACE (flagged LTEXT_FLAG_PREFORMATTED).
// This condition and the following one might handle keeping their
// widths constant, or might handle leading real space in the
// footnote number or footnote text...
// But as the are LTEXT_FLAG_PREFORMATTED, they might not yet
// collapse.
// This might need some rethinking.
```

Btw, I dropped the idea of using `float: left` I mentionned above mainly because with it, I couldn't ensure a space between the number and text.

**Comment by poire-z** [comment]
> the space between the number and the text in in-page footnotes is doubled comparing to popup footnotes

And which do you prefer ?

**Comment by hius07** [comment]
One space as in popup.

**Comment by hius07** [comment]
Good day,

This book shows no space after number in popup footnotes.
Besides, popup footnotes don't render the [emphasis] tag.
[Голсуорси Джон - В петле.fb2.zip](https://github.com/koreader/koreader/files/4906396/-.fb2.zip)

![Reader_2020-07-11_071545](https://user-images.githubusercontent.com/62179190/87218363-999e9000-c35a-11ea-97f0-1ea9a4bbac65.png)

**Comment by poire-z** [comment]
Unlike most of the others FB2, this one have the footnotes part stuck to each other in the source:
```html
<section id="n_6"><title><p>[6]</p></title><p><emphasis>¿¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿</emphasis> (1836¿1914) ¿ ¿¿¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿, ¿¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿ ¿¿¿¿ ¿ ¿¿¿¿ ¿¿¿¿¿-¿¿¿¿¿¿¿ ¿¿¿¿¿; ¿ 1895¿1903 . ¿¿¿¿¿¿¿ ¿¿¿¿ ¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿.</p>
</section>
```
(pardon my latin1 not showing cyrillic :)
Most others have their source a bit formatted:
```html
  <section id="n_10">
   <title>
    <p>10</p>
   </title>
   <p>¿¿¿¿¿¿¿¿ ¿ ¿¿. ¿¿¿¿¿¿. ¿¿¿¿¿.</p>
  </section>
```
So, when giving all that as-is to MuPDF, it gets some space between nodes and can show, once collapsed, at least a space between the number and the text - that it doesn't get in the first case.

I guess I could force adding one space after a node with  `display:run-in` when crengine gives the HTML to our frontend code that will give it to MuPDF.

For the italic missing, FB2 has some non-standard HTML tags (like `<emphasis>` that MuPDF does not know about, so it displays them unformatted.
So, we'd need to add formatting for these non-standard tags in:
https://github.com/koreader/koreader/blob/f4dad2fae8548b44a59a4ae6de9361673717ae77/frontend/ui/widget/footnotewidget.lua#L57-L111

One would need to find out the ones that need to be formatted differently.
https://github.com/koreader/crengine/blob/master/cr3gui/data/fb2.css
https://github.com/koreader/crengine/blob/59483446bc777b211cb23181bc8f6c60b94a1c1b/crengine/include/fb2def.h#L137-L231
https://github.com/koreader/crengine/blob/master/fb2props/FictionBook2.2.xsd

Quick glance and I see `emphasis` and `strikethrough` might need it.
Mind having a more serious look ?

**Comment by hius07** [comment]
> So, we'd need to add formatting for these non-standard tags in:

Will it be hardcoded that, say, [emphasis] is italic? Actually, it can be bold if one prefers to set it in fb2.css.

**Comment by poire-z** [comment]
It would be hardcoded (in footnotewidget.lua) in what we give to MuPDF for rendering popup footnotes. So, it should probably be the same as what we decided to be the default formatting for these tags in fb2.css.

**Comment by hius07** [comment]
> what we decided to be the default formatting for these tags in fb2.css

Can you please show me the thread discussing default fb2.css. May I say, it is not ideal.

**Comment by poire-z** [comment]
I don't think we ever discussed (or cared about :) fb2.css on the KOReader side. cf https://github.com/koreader/crengine/pull/194#issuecomment-392834289 . It's more a russian thing that we inherited from CoolReader.

The best you can get from us is the history of changes to this file: https://github.com/koreader/crengine/commits/master/cr3gui/data/fb2.css
and upstream has not changed it since we picked it: https://github.com/buggins/coolreader/commits/master/cr3gui/data/fb2.css
There are some issues opened there about fb2.css, like `https://github.com/buggins/coolreader/issues/137`.
But if anything has ever been discussed, it's probably on the Coolreader russian forum https://4pda.ru/forum/index.php?showtopic=191747

Anyway, as you're probably the person who cares the most about fb2.css, I think we can accept any change you'd like to make :)
Or open an issue here for discussion if you want feedback, and invite upstream russian FB2/FB3 lovers & devs.

**Comment by poire-z** [comment]
And for context, I guessed over the years that the history of all these CSS files might have been:
- Coolreader started with FB2 only, so fb2.css (and `crengine/include/fb2def.h` that defines all known elements, even those used for other formats).
- Then more formats were added, and html.css / epub.css were added, starting from a copy of fb2.css - so they carried many fb2 specific stuff over the years
- updates to these other formats CSS were may be (thoughlessly?) reported back into fb2.css
- and we got the mess we have (that we can't really get rid of, as some old books opened with these need them to get the original formatting so past highlights are not lost and can be migrated to the more stable DOM we have since the beginning of this year)

I mainly tried to make epub.css fixed and cleaner, and make obsolete all the others - except fb2.css that really needs something specific for its own non-HTML-standard element names like `<emphasis>` and its specific `<section>` stuff and footnotes.

**Comment by hius07** [comment]
Thank you very much, I'll examine all of that.

**Comment by hius07** [comment]
Seems to be one more issue with fb2.
fb2.css has instruction for [code] to be left-aligned. KOReader tries to justify it.
The second screenshot is from Coolreader for Win.
[1.fb2.zip](https://github.com/koreader/koreader/files/4908078/1.fb2.zip)

![1](https://user-images.githubusercontent.com/62179190/87239849-763c1980-c41c-11ea-8ddc-39023533c63f.png)
![2](https://user-images.githubusercontent.com/62179190/87239851-789e7380-c41c-11ea-8bdd-246c65db1e05.png)
I understand that this test book isn't usual (maybe incorrect) with that extra CR-LF's, anyway can you please look at it.

**Comment by poire-z** [comment]
Dunno how it should render, but telling you why it is that way:

In fb2.css, we have (stripped out some other element names from that):
```css
code, pre { font-family: "Courier New", "Courier", monospace; }
code, pre {
   white-space: pre;
   text-align: left;
   font-family: "Droid Sans Mono", "Liberation Mono", "DeJaVu Sans Mono", monospace;
   }
code {
   display: inline;
}
```
In your sample fb2 html:
```html
<body>
      <section>
          <p><code>one</code></p>
          <p><code>two two    <!-- there's a \n here -->
</code>
</p>
      </section>
</body>
```

CODE being `white-space: pre`, the \n is not ignored.
CODE being `display: inline`, it's just a part (the single part here) of the inline flow, and `text-align` does not apply to individual parts of the inline flow. So, it's the `text-align` of the real inline flow container, the P, that will apply: `justify`.
As `two two` is followed by a `\n`, `two two` is not the last line and so, the requested justification is done (while it is not done on the last line of a paragraph).

CoolReader does not have the `code { display: inline; }` , so it must handle it as block.
`code { display: inline; }` was added on our side, may be wrongly while fixing other xxx.css, with https://github.com/koreader/crengine/commit/97e2ea7e

Note that on our side (and probably picked by some dev code in upstream coolreader), we defaulted `code` to be inline in fb2def.h (because it is inline in HTML/EPUB).
So, if you want it block, you might have to explicitely set it `code { display: block; }` in fb2.css.

**Comment by hius07** [comment]
Thank you. got it.

**Comment by hius07** [comment]
Oh, sorry, I've changed `code { display: inline; }` to block in fb2.css, but still have it justified.

**Comment by hius07** [comment]
Well, I've added the same to User style tweaks and it is left-aligned finally.

**Comment by poire-z** [comment]
There's an other instance of it, below the one you changed (so it overrides your change), in the messy fb2.css, at the end of:
```css
b,strong,i,em,dfn,var,q,u,underline,del,s,strike,spacing,small,big,sub,sup,acronym,tt,sa mp,kbd,code {
   display: inline;
}
```

**Comment by poire-z** [comment]
> Well, I've added the same to User style tweaks and it is left-aligned finally.

I hope you're not giving up on cleaning up fb2.css :)
Now that we only use it with FB2 and FB3 (as we have a proper epub.css for other formats), you could really clean it up to only the element specified in the FB2 & FB3 specs.

**Comment by hius07** [comment]
> There's an other instance of it, below the one you changed

It's a trap :) Thanks!

**Comment by hius07** [comment]
I now refer to the `<v>` tag which is formatted in fb2.css as
`v { text-align: left; text-align-last: right; text-indent: 1em hanging }`
The screenshots are self explained.
[1.fb2.zip](https://github.com/koreader/koreader/files/4910536/1.fb2.zip)

![Reader_2020-07-13_064716](https://user-images.githubusercontent.com/62179190/87269704-a5af5c80-c4d6-11ea-9754-5df6c164f095.png)

This is from vlasovsoft CoolReader
![20200713064916](https://user-images.githubusercontent.com/62179190/87269721-b1028800-c4d6-11ea-8265-ccfe97edf82f.png)

The first and only short v-line should not be treated as last.

**Comment by poire-z** [comment]
> The first and only short v-line should not be treated as last.

May be it shouldn't for you styling FB2, but according to [the specs](https://drafts.csswg.org/css-text-3/#text-align-property) and Firefox, `text-align-last` applies to a single line (which is the first, but also the last). I "fixed" this recently in https://github.com/koreader/crengine/commit/5b28e2bd38a94e5b67aaf6c569be1e843072aedd - not yet picked by CoolReader.

I reckon it's strange and not what I would expect. But it's what Firefox & Chromium do for this:
```html
<html>
<body>
<p style="text-align: left; text-align-last: right">abcdefghij klmnopqrstuvwxyz</p>
<p style="text-align: left; text-align-last: right">abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz abcdefghij klmnopqrstuvwxyz</p>
</body>
</html>
```
<kbd>![image](https://user-images.githubusercontent.com/24273478/87277612-e90fc800-c4e2-11ea-9459-724fd5654ca3.png)</kbd>

So, I dunno how you could get what you want with CSS, no solution comes to my mind right now :( @Frenzie : any idea?
If really needed, we could add a private property value:  `text-align-last: -cr-right-if-not-single` ...

**Comment by hius07** [comment]
The `<v>` tag is for a single line in poems, so now I have all poems with short lines right-aligned, which is not nice.

**Comment by hius07** [comment]
I agree that generally `text-align-last` should be applied to a single line, but it is not good applying to the `<v>` tag.

**Comment by Frenzie** [comment]
Since you're dealing in elements a `v:last-child {blahblah}` should do what you want?

**Comment by poire-z** [comment]
Nope, what we have here is a single element: some long text in a single `<v>` (or `<p>`) that will be rendered as multiple lines. `:last-child` won't catch on text/line boxes.

**Comment by Frenzie** [comment]
Yes, last or only line is per spec afaict.

Here's a thought I had:

```html
<!DOCTYPE html>
<html>
<head>
<title>boo</title>
<style>
p {text-align-last:right}
p::first-line {text-align-last:left}
</style>
</head>
<body>
<p>
badger badger badger mushroom mushroom
badger badger badger mushroom mushroom
badger badger badger mushroom mushroom
badger badger badger mushroom mushroom
badger badger badger mushroom mushroom
badger badger badger mushroom mushroom
badger badger badger mushroom mushroom
badger badger badger mushroom mushroom
badger badger badger mushroom mushroom
</p>
<p>
badger badger badger mushroom mushroom
</p>
</body>
</html>
```

Works in Chromium (Vivaldi 3.1), not in Firefox 78.

![Screenshot_20200713_115744](https://user-images.githubusercontent.com/202757/87291816-3189ae80-c500-11ea-822f-55289098f598.png)

**Comment by poire-z** [comment]
That's a clever idea ! :)
Unfortunatly, we don't support `::first-line` - and it's harder to implement than the other pseudo elements/classes - wouldn't be that hard for your sample case, but for `p:first-line { font-weight: bold }` we wouldn't be able to use anymore the single full paragraph measuring we have (we would need to measure with bold to know where to cut, and re-measure the rest with the non-bold font).

Anyway, private named value - or go against the specs and get the previous behaviour (wouldn't mind that too much, because, to me, it felt more natural and expected :) and I have no idea in which context one would want `text-align-last` to behave as defined in the specs... and if anyone is really using it :)

**Comment by hius07** [comment]
`v::first-line {text-align-last:left}` doesn't work nor in fb2.fss neither in User style tweaks.

**Comment by Frenzie** [comment]
@hius07 Implicit in my thought above was that if you're implementing something new, it might make more sense to implement another part of the spec. ;-)

@poire-z Going against the spec is a potential option, but a bit dangerous. Perhaps something like `text-align-last: -cr-last-but-not-only`?

**Comment by Frenzie** [comment]
@poire-z Oh never mind, that's exactly what you already suggested earlier. ;-)

**Comment by Frenzie** [comment]
But I wrote it wrong. What I was thinking was clearly more along the lines of `-cr-test-align-last-but-not-only: right`. Either way, same principle.

**Comment by hius07** [comment]
Can `<v>` be a specific case, not equivalent to `<p>`?

**Comment by poire-z** [comment]
> What I was thinking was clearly more along the lines of `-cr-text-align-last-but-not-only: right`. 

Would waste a CSS property slot :) Cheaper to go with a named value for the existing one.

> Can `<v>` be a specific case

I'd rather not hardcode stuff, things should be driven by CSS. So, if we have one and you can use it in fb2.css, other people can use it for fixing other stuff in a user style tweaks.

We'll find some solution, let these `<v>` sit strangely in the meantime :)

**Comment by Frenzie** [comment]
> Would waste a CSS property slot :) Cheaper to go with a named value for the existing one.

Fair enough.

> I'd rather not hardcode stuff, things should be driven by CSS. So, if we have one and you can use it in fb2.css, other people can use it for fixing other stuff in a user style tweaks.

QFT :+1:

**Comment by hius07** [comment]
Back again with footnotes, now with the footnote number (link) in the text.
In fb2.css we have:
`a[type="note"] { vertical-align: super; font-size: 70%; text-decoration: none }`
Actually, the size of the link shown is 80% (it can he seen on the very first screenshot of this post, [9] is big).
Changing 70% to any value has no effect.
To manage the size I should state `font-size: 70% !important`.
In the meantime `vertical-align` and `text-decoration` can be managed without `!important`.

**Comment by poire-z** [comment]
Mhhh, no issue for me without `!important`  - if I set it to 50% :

```css
a { display: inline; text-decoration: underline; }
a[type="note"] { vertical-align: super; font-size: 50%; text-decoration: none }
```
<kbd>![image](https://user-images.githubusercontent.com/24273478/87438850-58bfa900-c5f0-11ea-9e3f-b9db0afed67d.png)</kbd>

Sure you don't have a style tweak enabled to ignore publisher font sizes ? Or some other changes to fb2.css ?

**Comment by hius07** [comment]
Thanks, have done more exploration and found the menu item which causes the issue:
Style tweaks - Misc - In-page footnotes - In-page EPUB footnotes (smaller)
(sic!)

**Comment by poire-z** [comment]
Oh, that's unfortunate :(
We have a conflict betweek the `EPUB in-page footnote smaller` tweak (enabled by default) that has `*[type~="note"]` to catch `<div class=note>123 Some footnote text</div>` - and FB2 footnote links having `type="note"`.

**Comment by poire-z** [comment]
OK, I should have fixed quite a few of the above issues in https://github.com/koreader/crengine/pull/357

**Comment by hius07** [comment]
Great, thanks! Waiting for the build to test.

**Comment by hius07** [comment]
Since fb2.css is touched, may I propose a minor clean up:
1) line 83 remove `, code, pre` (overrridden by 87)
2) line 135 remove `,code` (duplicate of 90)

**Comment by hius07** [comment]
Another idea is to add `hyphens: none` to the `<v>` tag.
The reason is obvious: unintendent hyphenation in poems hardens reading.

**Comment by poire-z** [comment]
OK.

> line 135 remove ,code (duplicate of 90)

So still letting `code` be `inline` ? Didn't you think above it should be `code { display: block; }` ?

**Comment by hius07** [comment]
I understand and agree with your explanation in https://github.com/koreader/koreader/issues/6344#issuecomment-657188503.
My example of the book is unusual and not perfectly formatted.

**Comment by poire-z** [comment]
OK, right, I checked in https://github.com/koreader/crengine/blob/master/fb2props/FictionBook2.2.xsd and CODE is some markup tag like STRONG, EMPHASIS.. so it should stay inline.

**Comment by poire-z** [comment]
Done there, have a quick look I did it allright.
(I think fb2.css would still deserve more cleanup :)

**Comment by hius07** [comment]
Thank you, I will be reviewing it periodically.
I don't know if it affects anything: a typo extra space in `sa mp`, line 131 in new
https://github.com/koreader/crengine/pull/357/commits/8a4cee335f9c4354301a14119b4089a2fd953280

**Comment by poire-z** [comment]
I had fixed that typo in epub.css, so ok, let's have it fixed in fb2.css too.

**Comment by Frenzie** [comment]
I was searching the archives a bit and came across this discussion btw https://lists.w3.org/Archives/Public/www-style/2014Jun/0167.html

(doesn't really add anything to ours here, besides touching on some of the same stuff)

**Comment by mergen3107** [comment]
Guys, I have some problems with footnote text size after 2020.07 update.
I use a custom fb2.css which is based on original fb2.css (by the way, [these](https://github.com/koreader/crengine/pull/357/files) changes did not overwrite in fb2.css after the update, I added them manually).

Regarding notes, I have this:
```
body[name="notes"] section title { display: run-in; text-align: left; font-size: 100%; font-weight: bold; page-break-before: auto; page-break-inside: auto; page-break-after: auto; }
body[name="notes"] section title p { display: inline }
```
which gives me notes like this:
![Reader_2020-07-20_193315](https://user-images.githubusercontent.com/6279855/87997063-0d127100-cac2-11ea-9529-ada32e52e6ee.png)
The option about keeping font size of 1rem in Styles - Misc - fb2 footnotes is off.
The notes page in the end of the book looks like this:
![Reader_2020-07-20_193330](https://user-images.githubusercontent.com/6279855/87997125-36cb9800-cac2-11ea-8ad0-2575342aae5a.png)


Now, if I just comment out everything about "notes", it becomes like this:
![Reader_2020-07-20_193450](https://user-images.githubusercontent.com/6279855/87997163-4f3bb280-cac2-11ea-9d40-85d8b1a5c023.png)

and each footnote is on a separate page in the end of book:
![Reader_2020-07-20_193506](https://user-images.githubusercontent.com/6279855/87997188-5d89ce80-cac2-11ea-8164-f60a1609443e.png)

1) My confusion is that I don't know how to tweak footnotes font size. Before that there was :
```
body[name="notes"] p { font-size: 67% !important; }
```
But now it makes not 67%, but even smaller font size. I commented it out. Moreover, enabling it makes larger line spacing in the footnote as well as in the notes body in the end of book. I'll attach some screenshots later.

2) By any chance, is it possible to make spacing after each notes, but only for only in the end of book? It would make it easier to read.

**Comment by poire-z** [comment]
> by the way, these changes did not overwrite in fb2.css after the update, I added them manually

It was bumped only last evening, so it's not yet in any nightly/release - but it will be in today's nightly, in one or hour two - can you recheck with it?

> Now, if I just comment out everything about "notes", it becomes like this:

The FB2 specs made that unfortunate decision to have the footnote number in a usually block element, and the text in another block element (they should be hung for that!):
```
SECTION > TITLE > P > number
          P > TEXT
```
So, if we don't do anything, that naturally displays as:
<kbd>![image](https://user-images.githubusercontent.com/24273478/88017483-0a496780-cb26-11ea-93f9-e074181d18fa.png)</kbd>
(TITLEs are centered by some other stuff in fb2.css)

The solution the original CoolReader author found to solve this (really really unsolvable case) is to introduce a very old (and forgotten and deprecated and never implemented in browsers) solution:
`display: run-in`
By setting it on `body[name="notes"] section title { display: run-in; }`, it becomes inline, but the next sibling too ! and they are both inline (and so, can happily comes on the same line, that's what we wanted), and their container becomes the block element:

<kbd>![image](https://user-images.githubusercontent.com/24273478/88017010-023cf800-cb25-11ea-8bf5-3cd09152dbf0.png)</kbd>

The nodes tagged with `~i` are inline, the ones with `~F` are "final" = block elements. Usually, `<P>` are `~F`.

> But now it makes not 67%, but even smaller font size
> `body[name="notes"] p { font-size: 67% !important; }`

That means the font of the paragraph block container is say 100%, but you only set `font-size: 67%` on some inner inline element: the paragraph line height (known as "the strut") is to be conserved according to the specs.
Because of the above solution, that P you target is no more a block container.

You need to target the SECTION element with a font-size, like we do in our style tweaks:
```css
body[name="notes"] > section {
    -cr-only-if: fb2-document;
        font-size: 0.75rem;
}
```

So, just forget your old CSS footnotes tweaks, and start from the ones we have and made fine these last weeks :)

> By any chance, is it possible to make spacing after each notes, but only for only in the end of book? It would make it easier to read.

I don't think so: the in-page footnotes are just slices and exact copies of the text at end of book. And there's no other container between each `body[name="notes"] > section` that you could put some margin on - and not have the margin of what becomes the in-page slice.

**Comment by mergen3107** [comment]
@poire-z 
Thank you very much! 
Yes, after today’s nightly update the fb2.css updated. Sorry, I didn’t look closely for timing of release. 

Thanks for the section tweak, now I see that font size changes appropriately!

**Comment by hius07** [comment]
Good evening, can you please see the screenshot below

![Reader_2020-07-26_213401](https://user-images.githubusercontent.com/62179190/88486777-cffa2300-cf88-11ea-8500-2c2877320d4d.png)

It shows endnote No. 172 in-page and in popup simultaneously.
We can see that the number of the in-page endnote is bigger.

**Comment by hius07** [comment]
Just for information one more screenshot with both footnotes and endnotes. 53 is a footnote.
![Reader_2020-07-26_204838](https://user-images.githubusercontent.com/62179190/88487084-de493e80-cf8a-11ea-9c72-dee4c0965d8d.png)

**Comment by poire-z** [comment]
When trying to find out if it's a fb2.css or a style tweak issue: disable all style tweaks, and go look how it looks at end of book.
Doing that, you see it's already bigger there, so fb2.css issue.
And this is the reason:
https://github.com/koreader/crengine/blob/28ebac45f7b1dc196836fbc86d6cba3424f14c6b/cr3gui/data/fb2.css#L61-L62
Looks like it can be solved with:
```diff
--- a/cr3gui/data/fb2.css
+++ b/cr3gui/data/fb2.css
@@ -104,6 +104,7 @@ body[name="notes"] section title,
 body[name="comments"] section title {
     display: run-in;   /* technical trick to have the footnote number inline with the followup text */
     font-weight: bold;
+    font-size: 100%; /* counteract default of 110% with regular title */
     text-align: start; /* counteract default of center with regular title */
     page-break-before: auto; /* counteract default of always with regular title */
     page-break-inside: auto;
```
Can you confirm this fixes that?

**Comment by hius07** [comment]
It fixes, thank you! And now I see that footnotes numbers were bigger too.
Now they all are OK.

**Comment by poire-z** [comment]
fb2.css related:
Upstream is fixing their https://github.com/buggins/coolreader/issues/137 with https://github.com/buggins/coolreader/pull/142/commits/ccac737f6899d8ed7e113c5d358febc293267308
I guess we should too.

Dunno if you have FB2 books with tables, but please have a look and decide how we want them.
(I'm not sure `border-style: ridge; border-color: #888;` is the best with eInk - gray colors usually cause a double refresh on my GloHD, some kind of distracting animation that I dislike :) dunno about `ridge`, I usually prefer `solid`.)

**Comment by hius07** [comment]
Sorry, I am not very experienced in looking into tables in my books.
In any case I am not a big fan of extra decoration, so black and solid would be my choice.

**Comment by mergen3107** [comment]
Yeah, the same here. I haven’t seen a table in my fb2 books yet, but for the sake of a table being just a well-structured text, I’d go with decoration as simple as possible

**Comment by virxkane** [comment]
About buggins/coolreader@ccac737: I just restored the legacy table view style (which was in the old CoolReader). May be we must create separate fb2.css for e-ink devices, for example fb2-eink.css (and corresponding for epub, etc formats). I'm not sure, maybe the koreader is only for e-ink devices, so this is not a actual idea?

**Comment by poire-z** [comment]
I guess KOReader is mostly used on eInk devices, but we build it for Android (so, it can run on color phones and tablets), desktop linux and MacOS.
I don't think we need to bother with different eink/non-eink css. We don't have much color anyway in our epub.css (and I seem to be the only one complaining about grey refreshes/animations :/):
- table headers & captions (which are rare), which are ok having grey background I guess
- `<HR>`, which are small
- `<A>` text, which we set grey (nobody ever complained about it - but I hate it, so I set it to be black & bold in my own epub.css :)

For fb2.css, `<A>` are black, and table headers & captions have the same background colors as in epub.css
Anyway, I'll probably set table borders to be solid black as KOReader does not have many FB2 users possibly used to the old ridge rendering (and personally, I find "ridge" a bit too much 1990s and early HTML :)
![border-styles](https://user-images.githubusercontent.com/24273478/89155654-8ce01700-d569-11ea-87af-32a3f982c1a3.png)

**Comment by Frenzie** [comment]
I would agree that ridge isn't the best looking option. I generally prefer a simple solid or nothing.

**Comment by hius07** [comment]
I wonder waht are the reasons of appearing of extra empty line between the comment (endnote) **7** and footnote **1**.
There are no empty lines in the text.

![1](https://user-images.githubusercontent.com/62179190/89978588-83943000-dc76-11ea-80c8-c2549241b5f6.png)

![2](https://user-images.githubusercontent.com/62179190/89978591-85f68a00-dc76-11ea-8368-7b0e990b5d8c.png)

![3](https://user-images.githubusercontent.com/62179190/89978593-87c04d80-dc76-11ea-8229-160d456e6a40.png)

![4](https://user-images.githubusercontent.com/62179190/89978598-898a1100-dc76-11ea-97ae-6edbb5ddf0c1.png)

**Comment by poire-z** [comment]
It's probably not some empty line of text, but some margin - precisely some collapsed margin from previous content.
I sometimes get that with some EPUBs in-page footnotes too.

Each footnote is a slice of the document, the part that contain that SECTION:
`body[name="notes"] section {-cr-hint: footnote-inpage;}`
Normally, all SECTIONs are similar, and should have no margin.
Except the first one, that may get the top margin of its container, or even bottom margin from a previous node - which is probably what happens here: the first footnote gets the bottom margin of the title in bold.
(You can switch rendering mode to "legacy", that doesn't do margin collapsing, to see if you don't get them.)

This margin collapsing is per-specs (https://github.com/koreader/crengine/pull/299).
In the implementation, we have to carry all margins met, to finally put it on the first element with content (so, the first footnote/SECTION here).
And this has this side effect with in-page footnotes, as we cut slices... we get the accumulated margin on the first footnote.

I have some patch (date Feb 13th) where I tried to fix that, that I did not follow up, probably because it caused some other issues:
```diff
--- a/crengine/src/lvrend.cpp
+++ b/crengine/src/lvrend.cpp
@@ -6561,8 +6561,15 @@ void renderBlockElementEnhanced( FlowState * flow, ldomNode * enode, int x, int

                 // Enter footnote body only after padding, to get rid of it
                 // and have lean in-page footnotes
-                if ( isFootNoteBody )
+                if ( isFootNoteBody ) {
+                    // If no padding were added, add an explicite 0-padding so that
+                    // any accumulated vertical margin is pushed here, and not
+                    // part of the footnote
+                    if (padding_top==0) {
+                        flow->addContentLine(0, RN_SPLIT_AFTER_AVOID, 0, true);
+                    }
                     flow->getPageContext()->enterFootNote( footnoteId );
+                }

                 // recurse all sub-blocks for blocks
                 int cnt = enode->getChildCount();
```

I can't get what you got with the FB2s I got from you.
Are you using our FB2 in-page footnotes style tweaks, or else ? They have `margin: 0 !important`, may be to counteract that?
If there's some specifc styling in that FB2, may be it can be solved by some additional styling in fb2.css, a bit like what I tried to do with the above patch. Something like:
```css
body[name="comments"] > section:first-of-type,
body[name="notes"] > section:first-of-type {
  padding-top: 1px;
}
     /* or */
body[name="comments"] > title,
body[name="notes"] > title {
  margin-bottom: 0;
}
```

**Comment by hius07** [comment]
Thank you, I'll play with some settings. Here is the book if you are interested to try.
[Golsuorsi_Forsayty-2.fb2.zip](https://github.com/koreader/koreader/files/5060995/Golsuorsi_Forsayty-2.fb2.zip)

**Comment by poire-z** [comment]
OK, I got it - and also with your other FB2 documents (somehow, I had all of them with render mode legacy :) 
Anyway, unless I got some day a better solution to avoid that, it can be solved by tweaking our style tweaks (a bit ugly, but well...)
```diff
--- a/frontend/ui/data/css_tweaks.lua
+++ b/frontend/ui/data/css_tweaks.lua
@@ -654,37 +654,46 @@ Show FB2 footnote text at the bottom of pages that contain links to them.]]),
                     -- match in any other kind of document
                     css = [[
 body[name="notes"] section {
     -cr-only-if: fb2-document;
         -cr-hint: footnote-inpage;
         margin: 0 !important;
 }
 body[name="notes"] > section {
     -cr-only-if: fb2-document;
         font-size: 0.75rem;
+}
+body[name="notes"] > title {
+    margin-bottom: 0;      /* avoid title bottom margin from collapsing into */
+    padding-bottom: 0.5em; /* the first footnote by by substituting it with padding */
 }
                     ]],
                 },
```

**Comment by hius07** [comment]
Can we avoid showing footnote number as `title` (bold and increased) when it is in the title?
![Reader_2020-08-21_055428](https://user-images.githubusercontent.com/62179190/90847706-cb504100-e373-11ea-8a91-96d2cc024a70.png)

**Comment by hius07** [comment]
I mean, I can do it via User style tweaks, but shouldn't it be by default?

**Comment by poire-z** [comment]
> I can do it via User style tweaks

How did you do it, what tweak did you came up with?

> but shouldn't it be by default?

For EPUB, I'd say it shouldn't (we respect cascading style sheets as the publisher might expect it).
For FB2, which is rarely styled, it's up to you readers (and how not ugly your style tweak can be :)

**Comment by hius07** [comment]
> How did you do it, what tweak did you came up with?

`a[type="note"] { font-weight: normal; }` - worked for title.
Didn't try `font-size`.

**Comment by poire-z** [comment]
We already have:
`a[type="note"] { vertical-align: super; font-size: 70%; text-decoration: none }`
I guess you also want `font-style: normal` to remove italic.
70% is a % of the container font size. If you want a fixed size relative to the default font size, you can use `font-size: 0.7rem` (rem = root em = root font size). Dunno how that will play with the vertical alignment (if you reduce the `a` font size, it may get lower, I don't remember if it uses the container font size or the contained font size to compute the vertical shift. Might also try `vertical-align: text-top` or `top` if needed).

**Comment by hius07** [comment]
Am I right that by default a nested (cascading) element should always inherit style of its parent strictly without exclusions?

**Comment by poire-z** [comment]
Half right.
It inherits only inheritable styles. https://stackoverflow.com/questions/5612302/which-css-properties-are-inherited
`font-weight`, `font-style` are inheritable.
`font-size` is not (otherwise, with `<span style="font-size: 70%"><span><span>test`, you'll get the text with a font size of 0.7 x 0.7 x 0.7 x the initial font size).

**Comment by hius07** [comment]
OK, thank you for your comments.
